packages/hydrogen/src/cart/optimistic/useOptimisticCart.tsx

```diff
@@ -0,0 +1,128 @@
+import {useFetchers} from '@remix-run/react';
+import type {cartGetDefault} from '../queries/cartGetDefault';
+import {CartForm} from '../CartForm';
+import type {
+  Cart,
+  CartLine,
+  ProductVariant,
+} from '@shopify/hydrogen-react/storefront-api-types';
+import {
+  getOptimisticLineId,
+  isOptimisticLineId,
+} from './optimistic-cart.helper';
+
+export type OptimisticCart<T = ReturnType<typeof cartGetDefault>> = T & {
+  isOptimistic?: boolean;
+  lines: {nodes: Array<CartLine & {isOptimistic?: boolean}>};
+};
+
+/**
+ * @param cart The cart object from `context.cart.get()` returned by a server loader.
+ *
+ * @returns A new cart object augmented with optimistic state. Each cart line item that is optimistically added includes an `isOptimistic` property. Also if the cart has _any_ optimistic state, a root property `isOptimistic` will be set to `true`.
+ */
+export function useOptimisticCart<
+  DefaultCart = ReturnType<typeof cartGetDefault>,
+>(cart: DefaultCart): OptimisticCart<DefaultCart> {
+  const fetchers = useFetchers();
+
+  if (!fetchers || !fetchers.length) return cart as OptimisticCart<DefaultCart>;
+
+  let isOptimistic = false;
+
+  const optimisticCart = cart
+    ? (structuredClone(cart) as OptimisticCart<DefaultCart>)
+    : ({lines: {nodes: []}} as OptimisticCart<DefaultCart>);
+
+  const cartLines = optimisticCart.lines.nodes;
+
+  for (const {formData} of fetchers) {
+    if (formData) {
+      const cartFormData = CartForm.getFormInput(formData);
+
+      if (cartFormData.action === CartForm.ACTIONS.LinesAdd) {
+        for (const input of cartFormData.inputs.lines) {
+          if (!input.selectedVariant) {
+            console.error(
+              'No selected variant was passed in the cart action. Make sure to pass the selected variant if you want to use an optimistic cart',
+            );
+            continue;
+          }
+
+          const existingLine = cartLines.find(
+            (line) =>
+              line.merchandise.id ===
+              (input.selectedVariant as ProductVariant)?.id,
+          );
+
+          isOptimistic = true;
+
+          if (existingLine) {
+            existingLine.quantity = (existingLine.quantity || 1) + 1;
```

Comment:
Are we making the assumption here that merchants will only allow adding one item at a time? Some merchants will provide a quantity selector on their PDPs